src/Razor/src/Microsoft.CodeAnalysis.Remote.Razor/RazorServiceBase.cs

```diff
@@ -1,29 +1,71 @@
 ﻿// Copyright (c) .NET Foundation. All rights reserved.
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
+#nullable enable
+
 using System;
 using System.Collections.Generic;
 using System.IO;
+using System.IO.Pipelines;
 using System.Threading;
 using System.Threading.Tasks;
 using Microsoft.AspNetCore.Razor.Language;
 using Microsoft.CodeAnalysis.Razor;
 using Microsoft.CodeAnalysis.Razor.ProjectSystem;
-using Newtonsoft.Json;
+using Microsoft.ServiceHub.Framework;
+using Microsoft.ServiceHub.Framework.Services;
+using Nerdbank.Streams;
 
 namespace Microsoft.CodeAnalysis.Remote.Razor
 {
-    internal abstract class RazorServiceBase : ServiceHubServiceBase
+    internal abstract class RazorServiceBase : IDisposable
     {
-        public RazorServiceBase(Stream stream, IServiceProvider serviceProvider)
-            : base(serviceProvider, stream, GetRazorConverters())
+        internal abstract class FactoryBase<TService> : IServiceHubServiceFactory
```

Comment:
This is copied from Roslyn and is designed to make it easy to define descriptors for multiple services. it can be simplified if Razor continues to have just a single ISB service.

Reply:
We would need to wrap it in external access wrappers, making it ugly and complicated.